spring-boot-project/spring-boot/src/main/java/org/springframework/boot/jdbc/EmbeddedDatabaseConnection.java

```diff
@@ -106,7 +107,7 @@ public String getUrl() {
 	 */
 	public String getUrl(String databaseName) {
 		Assert.hasText(databaseName, "DatabaseName must not be null.");
-		return String.format(this.url, databaseName);
+		return this.url != null ? String.format(this.url, databaseName) : null;
```

Comment:
That wasn't the intention as raised on #8690 - The idea is to avoid a NPE by an abuse of `DataSourceProperties` (as explained in the comments). I'd rather have an `IllegalStateException` here that states the url should not be null or something.

Thoughts?

Reply:
I think the result for the user would basically the same. 
With this implementation they get an ``DataSourceBeanCreationException `` with the message _Cannot determine embedded database url_
If I change it to your proposal they get an ``IllegalStateException``with a message that states the url should not be null or something.

Reply:
The difference is that you give a perfectly valid exception message for a use case that isn't. I don't know if you've picked up that issue or if you faced it yourself but the issue as described doesn't result from a legit use case. I think we should then treat it as such.

Reply:
But it is a legit use case to define DataSourceProperties without a url - at least in the current implementation. Because then this fallback to the embedded database takes place (in my opinion it would be better to not have this magic at all but it’s there already so I don’t think it can/should be changed and is definitely not in scope of this issue.)
The error case from the issue is that no url is defined *and* no embedded database is available. 
I just think the exception should be thrown from the DataSourceProperties and not from the EmbeddedDatabaseConnection because if you don’t even add an embedded database, why should you get an error that originates in an EmbeddedDatabaseConnection class?
But maybe I misunderstand this completely...?

Reply:
> But maybe I misunderstand this completely...?
 
Or I do. Let me run that thing concretely, thanks for the feedback.